### PR TITLE
#77: property for checking the order

### PR DESCRIPTION
**Pull request by jgheerardyn**

### CODE REVIEW

File: foundations/src/main/java/io/yields/math/framework/property/SortedProperty.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    public boolean isValid(Collection<T> value, FunctionalContext<Collection<T>> context) {
+        Validate.notEmpty(value, "An empty collection cannot be ordered.");
+        int sign = increasing ? 1 : -1;
+        Iterator<T> sorted = value.stream().sorted((t1, t2) -> sign * t1.compareTo(t2)).iterator();
+        Iterator<T> original = value.iterator();
+        while (original.hasNext()) {
+            if (original.next().compareTo(sorted.next()) != 0) {
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/property/SortedProperty.java]: Avoid iterators.
Instead prefer sth like this:

List<T> sorted = value.stream().sorted(...).collect(toList());
IntStream range = IntStream.range(0, value.size());
range.findFirst(index -> value.get(index).equals(sorted.get(index)) != 0).map(invalidEntry => false).orElse(true)

**Comment by rombautt** [comment]
Reviewed

File: foundations/src/main/java/io/yields/math/framework/property/SortedProperty.java (hunk 1 above)
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/property/SortedProperty.java]: Just out of curiosity, why is it better to avoid iterators here?

Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/property/SortedProperty.java]: This code is not bad per se, but an iterator has some drawbacks:

- it's statefull
- it's can only be iterated once, single-threaded

Now we have java 8, I prefer to use streams; we could easily speed up things by making the streams parallel, whereas iterators remain single-threaded (as not thread-safe).

**Pull request merged**